backend/btrixcloud/zip.py

```diff
@@ -0,0 +1,147 @@
+"""
+Methods for interacting with zip/WACZ files
+"""
+import io
+import json
+import os
+import struct
+import zipfile
+import zlib
+
+from fastapi import HTTPException
+
+
+# ============================================================================
+EOCD_RECORD_SIZE = 22
+ZIP64_EOCD_RECORD_SIZE = 56
+ZIP64_EOCD_LOCATOR_SIZE = 20
+
+MAX_STANDARD_ZIP_SIZE = 4_294_967_295
+
+
+# ============================================================================
+async def extract_and_parse_log_file(client, bucket, key, log_zipinfo, cd_start):
+    """Return parsed JSON from extracted and uncompressed log"""
+    # pylint: disable=too-many-locals
+    file_head = await fetch(
+        client, bucket, key, cd_start + log_zipinfo.header_offset + 26, 4
+    )
+    name_len = parse_little_endian_to_int(file_head[0:2])
+    extra_len = parse_little_endian_to_int(file_head[2:4])
+
+    content = await fetch(
+        client,
+        bucket,
+        key,
+        cd_start + log_zipinfo.header_offset + 30 + name_len + extra_len,
+        log_zipinfo.compress_size,
+    )
+
+    if log_zipinfo.compress_type == zipfile.ZIP_DEFLATED:
+        uncompressed_content = zlib.decompressobj(-zlib.MAX_WBITS).decompress(content)
+    else:
+        uncompressed_content = content
+
+    content_length = len(uncompressed_content)
+    if not log_zipinfo.file_size == content_length:
+        # pylint: disable=line-too-long
+        detail = f"Error extracting log file {log_zipinfo.filename} from WACZ {os.path.basename(key)}."
+        detail += f" Expected {log_zipinfo.file_size} bytes uncompressed but found {content_length}"
+        print(detail, flush=True)
+        raise HTTPException(status_code=500, detail=detail)
+
+    parsed_log_lines = []
+
+    for json_line in uncompressed_content.decode("utf-8").split("\n"):
```

Comment:
It looks like the way this is setup is it's reading the entire log file into memory, than splitting by newlines.
That may be fine for first pass, though logs can get quite large. We probably want to actually make this stream..

In looking up ways to do this, I actually found this library which might be of help / simplify things:
https://pypi.org/project/smart-open/

Reply:
See above comment instead, should be fairly simple to adjust.